### PR TITLE
Move image loading out of constructors and into static methods

### PR DESCRIPTION
**Pull request by tocsoft**
### Prerequisites

- [x] I have written a descriptive pull-request title
- [x] I have verified that there are no overlapping [pull-requests](https://github.com/JimBobSquarePants/ImageSharp/pulls) open
- [x] I have verified that I am following matches the existing coding patterns and practise as demonstrated in the repository. These follow strict Stylecop rules :cop:.
- [x] I have provided test coverage for my change (where applicable)

### Description
I've refactored Image loading (from stream, byte[] and files) to move the logic out of the `Image<TColor>` constructor and into static methods.

In the process I was able to clean up a a chunk of the Image data loading internals and additionally a shifted the responsibility of constructing the image object into the `IImageEncoder`s instead of user code. This allowed me to remove the `InitPixels(..)` method as they will always be initialised on construction.

Additionally I removed the optional parameters in from the constructors and now they are using proper overloads.

### CODE REVIEW

**Comment by antonfirsov** [comment]
Makes sense to me :+1: 

We can even consider the following: With static Load() methods we are capable to pass an `IImageDecoder` or an `IImageFormat` explicitly:
- Users might restrict their load call to a specific decoder
- Better encapsulation for tests: the decoder selection logic won't be executed when we are testing a specific decoder.

**Comment by tocsoft** [comment]
This update will actually allow you to do this for testing 
```c#
Image<Color> image =new JpegDecoder().Decode<Color>(stream, options);
```

**Comment by antonfirsov** [comment]
Hmm ... you're right, but maybe it's worth to define an `Image.Load(stream, decoder, options)` wrapper for the sake of API consistency.

**Comment by tocsoft** [comment]
Overloads added
 You can now call 

```c#
byte[] bytes = .......;
Image img = Image.Load(bytes, new PngDecoder());
```

**Comment by JimBobSquarePants** [comment]
@tocsoft I'll do a review of this asap. I'm reading up on chromatic adaption just now so I can write unit tests.

**Comment by tocsoft** [comment]
I just realised that when returning an `Image` as aposed to `Image<TColor>` at the moment I'm duplicating the pixel data... small change to come which should resolve them.

**Comment by tocsoft** [comment]
@JimBobSquarePants fixed the problem I spotted

**Comment by tocsoft** [comment]
Is everyone now happy with this?

File: src/ImageSharp/Image.Create.cs [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+    using System.Diagnostics;
+    using System.IO;
+
+    using Formats;
+
+    /// <summary>
+    /// Represents an image. Each pixel is a made up four 8-bit components red, green, blue, and alpha
+    /// packed into a single unsigned integer value.
+    /// </summary>
+    public sealed partial class Image
```
Comment (JimBobSquarePants) [inline src/ImageSharp/Image.Create.cs]: This feels odd. We're extending the non-generic Image class with Image<TColor> methods. I get that it works, it just feels odd.

File: src/ImageSharp/Image.Decode.cs [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+    using System.Text;
+    using Formats;
+
+    /// <summary>
+    /// Represents an image. Each pixel is a made up four 8-bit components red, green, blue, and alpha
+    /// packed into a single unsigned integer value.
+    /// </summary>
+    public sealed partial class Image
+    {
+        private static IImageFormat DiscoverFormat(Stream stream, Configuration config)
```
Comment (JimBobSquarePants) [inline src/ImageSharp/Image.Decode.cs]: XML docs please

File: src/ImageSharp/Image.FromFile.cs [hunk 3]
```diff
@@ -0,0 +1,218 @@
+﻿// <copyright file="Image.FromStream.cs" company="James Jackson-South">
+// Copyright (c) James Jackson-South and contributors.
+// Licensed under the Apache License, Version 2.0.
+// </copyright>
+
+namespace ImageSharp
+{
+    using System;
```
Comment (JimBobSquarePants) [inline src/ImageSharp/Image.FromFile.cs]: Using's should be inside `#if`

File: src/ImageSharp/Image.FromStream.cs [hunk 4]
```diff
@@ -0,0 +1,248 @@
+﻿// <copyright file="Image.FromStream.cs" company="James Jackson-South">
+// Copyright (c) James Jackson-South and contributors.
+// Licensed under the Apache License, Version 2.0.
+// </copyright>
+
+namespace ImageSharp
+{
+    using System;
```
Comment (JimBobSquarePants) [inline src/ImageSharp/Image.FromStream.cs]: Unused references in here. Buffer, Diagnostics, Linq

File: src/ImageSharp/Image.FromStream.cs [hunk 5]
```diff
@@ -0,0 +1,248 @@
...
+        /// <exception cref="NotSupportedException">
+        /// Thrown if the stream is not readable nor seekable.
+        /// </exception>
+        /// <returns>The image</returns>
+        public static Image<TColor> Load<TColor>(Stream stream, IDecoderOptions options, Configuration config)
+            where TColor : struct, IPixel<TColor>
+        {
+            config = config ?? Configuration.Default;
+
+            Image<TColor> img = WithSeekableStream(stream, s =>
```
Comment (JimBobSquarePants) [inline src/ImageSharp/Image.FromStream.cs]: Can be. 

```
Image<TColor> img = WithSeekableStream(stream, s => Decode<TColor>(stream, options, config));
```

File: src/ImageSharp/Image.FromBytes.cs [hunk 6]
```diff
@@ -0,0 +1,216 @@
+﻿// <copyright file="Image.FromStream.cs" company="James Jackson-South">
+// Copyright (c) James Jackson-South and contributors.
+// Licensed under the Apache License, Version 2.0.
+// </copyright>
+
+namespace ImageSharp
+{
+    using System;
```
Comment (JimBobSquarePants) [inline src/ImageSharp/Image.FromBytes.cs]: Unused references in here. Buffer, Diagnostics, Linq, Text

File: src/ImageSharp/Image.cs [hunk 7]
```diff
@@ -219,7 +49,7 @@ public Image(byte[] bytes, IDecoderOptions options, Configuration configuration)
         /// </summary>
         /// <param name="other">The other image, where the clone should be made from.</param>
         /// <exception cref="System.ArgumentNullException"><paramref name="other"/> is null.</exception>
-        public Image(Image other)
+        internal Image(Image<Color> other)
```
Comment (JimBobSquarePants) [inline src/ImageSharp/Image.cs]: How do I now deep clone an image?

Comment (tocsoft) [inline src/ImageSharp/Image.cs]: not sure why I did that. 😕

File: src/ImageSharp/Image.FromBytes.cs (hunk 6 above)
Comment (tocsoft) [inline src/ImageSharp/Image.FromBytes.cs]: theses have been removed

File: src/ImageSharp/Image.FromStream.cs (hunk 5 above)
Comment (tocsoft) [inline src/ImageSharp/Image.FromStream.cs]: done

File: src/ImageSharp/Image.FromFile.cs (hunk 3 above)
Comment (tocsoft) [inline src/ImageSharp/Image.FromFile.cs]: done

File: src/ImageSharp/Image.Create.cs (hunk 1 above)
Comment (tocsoft) [inline src/ImageSharp/Image.Create.cs]: It felt the best from the public api point of view (felt it ended up being the most discoverable via intellisense).

Comment (JimBobSquarePants) [inline src/ImageSharp/Image.Create.cs]: Aye, Enumerable does the same.

**Comment by JimBobSquarePants** [comment]
I'm happy to have this merged now.

**Comment by dlemstra** [comment]
I rather have the copying of the metadata in the copy constructor. Don't really like the LoadFrom method because you can call it multiple times. But it makes no sense to me to have this option. Also leaving a minor comment about the argument order. My OCD says configuration should be first.

**Comment by tocsoft** [comment]
Had a play with the argument ordering and totally agree moved the config args to the start.

**Comment by tocsoft** [comment]
@dlemstra I've taken your feedback into account you happier with this now?

**Pull request merged**